src/callbacks/perturb_weights.cpp

```diff
@@ -0,0 +1,166 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/comm_impl.hpp"
+#include "lbann/callbacks/perturb_weights.hpp"
+#include "lbann/proto/proto_common.hpp"
+#include "lbann/utils/serialize.hpp"
+#include "lbann/weights/data_type_weights.hpp"
+
+#include <callbacks.pb.h>
+
+
+namespace lbann {
+namespace callback {
+
+perturb_weights::perturb_weights(std::string output_name,
+                           El::Int batch_interval)
+  : callback_base(batch_interval),
+    m_output_name(std::move(output_name)) {
+    m_output = nullptr;
+}
+
+perturb_weights::perturb_weights()
+  : perturb_weights("",0)
+{}
+
+template <class Archive>
+void perturb_weights::serialize(Archive & ar) {
+  ar(::cereal::make_nvp(
+       "BaseCallback",
+       ::cereal::base_class<callback_base>(this)),
+     CEREAL_NVP(m_output_name));
+}
+
+void perturb_weights::setup(model* m) {
+   for (auto* w : m->get_weights()) {
+      if(w->get_name() == m_output_name){
+        m_output = w;
+        break;
+      }
+   }
+    if (m_output == nullptr) {
+      LBANN_ERROR("Current implementation of callback \"", name(), "\" "
+                  "requires a weight object to perturb");
+    }
+}
+
+void perturb_weights::on_batch_begin(model* m) {
+  const auto& c = m->get_execution_context();
+  if (m_output != nullptr && 
+      c.get_step() % m_batch_interval == 0 &&
+      c.get_step() >= 0) {
+    perturbed(*m);
+  }
+}
+
+void perturb_weights::perturbed(model& m){
+
+  auto* comm = m.get_comm();
+
+  // Useful constants
+  constexpr DataType zero = 0;
+  constexpr DataType one = 1;
+  const auto range = 3;
+  const auto thres = 0.5;
+  const auto lower = 0.3;
+  const auto upper = 0.7;
+  
+  // RNG
+  auto& gen = get_generator();
+  std::normal_distribution<DataType> dist(zero, one);
+
+
+  for (auto* w : m.get_weights()) {
+    if (w == nullptr) {
+    	LBANN_ERROR("callback \"", name(), "\" "
+                  "got a weights pointer that is a null pointer");
+    }
+
+    // Check layer name
+    if(w->get_name() == m_output_name) {
+	auto& values = w->get_values();
+	auto& new_values = dynamic_cast<El::AbstractDistMatrix<DataType>&>(values);
+
+	auto& local_values = new_values.Matrix();
+	El::Matrix<DataType,El::Device::CPU> temp;
+	El::Copy(local_values, temp);
+
+	// Perturb weights on master process		
+	if (comm->am_trainer_master()) {
+		for (auto i = 0; i < range; i++){		
+			
+			// perturb				
+			auto val = temp.Get(i,0);
+			auto perturbed_val = val;
+			
+			if(val >= lower && val <= upper){	
+				if (dist(gen) > thres){
+					perturbed_val = val * 1.1;
+				}
+				else {
+					perturbed_val = val * 0.9;
+				}
+			}

```

Comment:
If I am understanding this correctly, you want a 50% chance of either increasing the weights by 1.1x or decreasing by 0.9x? In that case, the distribution should be uniform instead of normal.

This works, but it forces the weights to take discrete values. If we want continuous values, we might try something like:
```suggestion
			perturbed_val += dist(gen); // dist is a std::normal_distribution
			perturbed_val = std::min(std::max(perturbed_val, lower), upper);
```